### PR TITLE
Optimize header

### PR DESCRIPTION
**Pull request by michaelletzgus**
Idea: Stop notification from overlapping everything else:

* Optimize header, introduce header-center, place notifications in it
* Please ignore colors, will be removed in final version

Is this something worth finishing?

### CODE REVIEW

**Comment by MorrisJobke** [comment]
The notifications doesn't appear anymore when I execute: `OC.Notification.show('Hello - I\'m a notification')` - work fine on master. Could you check that?

**Comment by jancborchardt** [comment]
@michaelletzgus didn’t check yet but it’s a good initiative! cc @nextcloud/designers for review.

@michaelletzgus please make sure that it also works well on mobile (narrow screens) and maybe also check out https://github.com/nextcloud/server/issues/301 (proposal for non-blocking notifications)

**Comment by juliushaertl** [comment]
I like the idea, but the issue with blocking UI should be fixed when touching the code as well:

2017-03-28-181008_638x110_scrot

Are those notifications being removed after some time? I'm not sure but I think that was already in somewhere in the past?

**Comment by jancborchardt** [comment]
Just checked and this will definitely create issues with the apps being in the header. It’s strange to push the app list away just to show a notification. It’s basically the same issue of overlaying again, but even worse :\

**Comment by michaelletzgus** [comment]
The current version in my branch is "under construction". My current goal is to create a (modal?) overlay
in the screen's center when space is limited.
Problem: WHEN exactly is space limited?
And: Why not use an popup-overlay always?

**Comment by michaelletzgus** [comment]
The notification bar now collapses into a screen centered popup at width<680.
(The CSS currently contains a small design flaw: The popup overlay is a child of `header-center`, will be fixed)

Question: Should ALL notifications be shown in a popup - even with wide screen? Or should they stay in the header?

File: core/js/js.js [hunk 1]
```diff
@@ -1158,17 +1158,16 @@ OC.Notification={
...
-			var $closeButton = $('<a class="action close icon-close" href="#"></a>');
-			$closeButton.attr('alt', t('core', 'Dismiss'));
-			$row.append($closeButton);
-			$closeButton.one('click', function() {
-				self.hide($row);
-				return false;
-			});
-			$row.addClass('closeable');
-		}
+//		// add a close button
```
Comment (MorrisJobke) [inline core/js/js.js]: No need to double comment :D

File: core/templates/layout.user.php [hunk 2]
```diff
@@ -107,8 +104,20 @@
...
 						</div>
-					</div></nav>
+					</div>
+				</nav>
+
+				<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
+					$(document.createElement( "button" )).click(function(){ OC.Notification.show('Hello - I am a notification') }).appendTo($('#header-left')).html("B1");
+					$(document.createElement( "button" )).click(function(){ OC.Notification.showHtml('Hello - I <b>am</b> a very very <i>loooooong</i> <u>HTML noooooooooooootification</u>....................') }).appendTo($('#header-left')).html("B2");
+					$(document.createElement( "button" )).click(function(){ OC.Notification.showTemporary('temp. notification',{timeout:2}) }).appendTo($('#header-left')).html("B3");
+				</script>
```
Comment (MorrisJobke) [inline core/templates/layout.user.php]: Please remove debug code and just put it in the description. Everybody can then execute it themselves in the web dev console.

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
See comments

**Comment by michaelletzgus** [comment]
Of course it's not centered on screen because the free space in the header isn't.

**Comment by jancborchardt** [comment]
@michaelletzgus it would be a bit nicer to have a solution where the notification appears in a place which is consistent. It’s strange when it moves depending on other circumstances. Did you consider showing it below the header bar?

**Comment by michaelletzgus** [comment]
A consistent place would be nice - but below the header just moves the "invasive overlay problem" from the header a few pixels to the bottom.
I think a few questions should be answered before the redesign:

1. What is an important message?
2. Should important messages be invasive?
3. Are important very messages so important that they justify a modal overlay?
4. What are "notes", "hints" and "tips of the day"?

I think everything that can NOT be safely ignored (configuration problems, security problems, major (minor?) server updates, app updates are important and might be invasive/modal...

**Comment by jancborchardt** [comment]
> What is an important message?

A new version being available for example.

> Should important messages be invasive?

A bit, in the style that they are now, with a notification up top. They shouldn’t just be in the Notification app, should they @LukasReschke @nickvergessen?

> Are important very messages so important that they justify a modal overlay?

Definitely not.

> What are "notes", "hints" and "tips of the day"?

Anything like that should be displayed inline, not as a notification and especially not as a modal.

What do you think @nextcloud/designers?

**Comment by nickvergessen** [comment]
The header notifications are e.g. failed upload attempts via the web UI.
They should not be pushed to your desktop/mobile client, so not related to notifications app => *unsubscribing*

**Comment by MorrisJobke** [comment]
> A new version being available for example.

I would not say, that this specific case is so important. A new version is fine to have in the notifications area.

I would only like stuff in this direct notification on the top for stuff that just happened. But the more I think about use cases the more it feels wrong to have this here. It's just a generic and ugly approach to handle feedback for the user, but actually it shouldn't be done like it is now.

Some examples:

* undo user deletion: currently such an notification, but it should be shown where the user entry was before
* failed to share a file because of REASON: this should be shown in the sharing sidebar - just there where you clicked to create the share
* a new version is available: should be a real notification in our notification drop down and also be pushed to the clients, because this is an actual state of the server and not a state of the current session


The more I think about this, the more I think we should get rid of OC.Notifications and either use our notification center (for state of the instance, state of a user) or inline feedback/inline actions for stuff that just happened (like error handling and undo actions).

Opinions @nextcloud/designers ?

**Comment by jancborchardt** [comment]
@MorrisJobke yep, totally agree. As I said above:

> To be honest, we want to get rid of most notifications in apps because they are just used as an easy way out but are not very nicely designed. It’s better to do it with `.emptycontent` or inline.

And everything which doesn’t work via emptycontent or inline, is probably best in the Notifications app (like a new version being available).

**Comment by eppfel** [comment]
So, we should therefore collect occasions where this nasty notification is still used and step by step remove it by putting custom notifications close to the context.

Closing, as it is a fix for a symptom, not for a cause. But thanks for kickstarting the discussion @michaelletzgus

**Pull request closed without merging**